R/otp-route.R

```diff
@@ -0,0 +1,287 @@
+#'
+#' @export
+#'
+#' @detials
+#' This function returns a SF data.frame with one row for each leg of the jounrey
+#' (a leg is defined by a change in mode). For transit more than one route option may be returned
+#' and is indicated by the route_option column.
+#'
+#'
+#' Elevation
+#' OTP supports elevation data, and can return the elevation profile of the route if available.
+#' OTP returns the elevation profile separately from the XY coordinates, this means there is not
+#' direct match between the number of XY points and the number of Z points.  OTP also only returns
+#' the elevation profile for the first leg of the route (this appears to be a bug).
+#' As default the otp_plan function matches the elevation profile to the XY coordinates to return
+#' a SF linestring with XYZ coordinates. If you require a more detailed elevation profile,
+#' the full_elevation parameter will return a nested data.frame with three columns.
+#' first and second are returned from OTP, while distance is the cumulative distance along the
+#' route and is derived from First.
+#'
+otp_plan <- function(otpcon = NA,
+           fromPlace = NA,
+           toPlace = NA,
+           mode = "CAR",
+           date_time = Sys.time(),
+           arriveBy = FALSE,
+           maxWalkDistance = 800,
+           walkReluctance = 2,
+           transferPenalty = 0,
+           minTransferTime = 0,
+           full_elevation = FALSE)
+{
+  # Check Valid Inputs
+  if(!"otpconnect" %in% class(otpcon)){
+    message("otpcon is not a valid otpconnect object")
+    stop()
+  }
+  if(class(fromPlace) != "numeric" | length(fromPlace) != 2){
+    message("fromPlace is not a valid latitude, longitude pair")
+    stop()
+  }else{
+    if(fromPlace[1] <= 90 & fromPlace[1] >= -90 &  fromPlace[2] <= 180 & fromPlace[2] >= -180){
+      fromPlace <- paste(fromPlace, collapse = ",")
+    }else{
+      message("fromPlace coordinates excced valid values +/- 90 and +/- 180 degrees")
+      stop()
+    }
+
+  }
+  if(class(toPlace) != "numeric" | length(toPlace) != 2){
+    message("toPlace is not a valid latitude, longitude pair")
+    stop()
+  }else{
+    if(toPlace[1] <= 90 & toPlace[1] >= -90 &  toPlace[2] <= 180 & toPlace[2] >= -180){
+      toPlace <- paste(toPlace, collapse = ",")
+    }else{
+      message("toPlace coordinates excced valid values +/- 90 and +/- 180 degrees")
+      stop()
+    }
+
+  }
+  if(!all(mode %in% c("TRANSIT","WALK","BICYCLE","CAR","BUS","RAIL"))){
+    message("mode is not a valid, can be a character vector of any of these values TRANSIT, WALK, BICYCLE, CAR, BUS, RAIL")
+    stop()
+  }else{
+    mode <- paste(mode, collapse = ",")
+  }
+  if(!all(class(date_time) %in% c("POSIXct","POSIXt") )){
+    message("date_time is not a valid, must be object of class POSIXct POSIXt")
+    stop()
+  }else{
+    date <- substr(as.character(date_time),1,10)
+    time <- substr(as.character(date_time),12,16)
+  }
```

Comment:
@mem248 Having problems get TRANSIT (or BUS) route returned for a specific date. For example, the following on OTP with my Manchester tutorial data, returns trip is not possible error but a valid route is returned using the web GUI, where otpdate is `"2017-10-10 12:15:00 BST"`:

```
otp_plan(otpcon, from = c(53.56172, -2.29202), to = c(53.40892, -2.16637), mode = "BUS", full_elevation = FALSE, date_time = otpdate)
```

This may be due to the date being in the format YYYY-MM-DD when it's extracted from the POSIXct object, while OTP requires MM-DD-YYYY? On that point, I wonder if we really want to require users to provide date and time in POSIXct? It's not that user friendly, and many won't know how to create the object. I had to Google it to remind myself. Think I'd rather provide a basic text format option (or the option of either). What do you think?

Reply:
I know that POSIXct is a little awkward, but it does ensure that we get properly formatted dates. Otherwise, there would be a wide range of formats that would need to be supported or checked.


OTP doesn't seem to support routing in the past, so I would not expect a 2017 date to work

Reply:
@mem48 OTP will route based on the schedules provided. If they are in the past then it will route on them.  The schedules in my test instance covered that period and worked in the GUI. But the format must be MM-DD-YYYY in the API request. You can see that in the GUI URL and also see the effect of altering it to YYYY-MM-DD.

Reply:
OTP only supports one format though. So as long as the documentation specified:

time as "13:00:00" and date as "MM-DD-YYYY" we would only have to check for that?

Reply:
Obviously the date of travel specified in the request must match the transit service validity period (as provided in calendar.txt in the GTFS feed).

Reply:
Actually this is one of the great features of OTP (compared to using an online trip planner that will only ever have the current schedules) , you can provide it with old or future data for research. E.g. look at how accessibility has changed due to reduction in bus services over the past 25 years ...

Reply:
I've adjusted the date and time formatting passed to the API, which I think is part of the problem. I can not get historical routing, as far back as the GTFS file goes. Also, are you using the GTFS in the package or your tutorial one? I made some modifications to the one in the package.

I still would prefer the have POSIXct inputs, but perhaps the vignettes can highlight an easy way to make dates e.g. `date_time = lubridate::ymd_hms("2018-04-07 12:15:00")`